### PR TITLE
Single file data loader

### PR DESCRIPTION
**Pull request by sahanediriweera**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -2,15 +2,35 @@
...
+
+def count_csv_rows(csv_file):
+    row_count = 0
+    with open(csv_file, 'r') as file:
+        reader = csv.reader(file)
+        for row in reader:
+            row_count += 1
+    return row_count
+
+filename = 'data_1000.csv'
```
Comment (chxru) [inline app.py]: You can't run the program if you don't have the CSV file. Gonna be a bad experience in fresh runs. Re write this flow

1. Check if the csv exists
2. If file exists, load it
3. Else, generate the csv, save it in disk and load it

File: app.py [hunk 2]
```diff
@@ -2,15 +2,35 @@
...
+            data.append([float(x) for x in row])
+    return np.array(data)
+
+def count_csv_rows(csv_file):
+    row_count = 0
+    with open(csv_file, 'r') as file:
+        reader = csv.reader(file)
+        for row in reader:
+            row_count += 1
+    return row_count
```
Comment (chxru) [inline app.py]: Might want to look into this
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.count.html#pandas-dataframe-count

File: app.py [hunk 3]
```diff
@@ -20,8 +40,10 @@
 # split data into chunks
 N_per_process = N // size
 
-# calculating kmeans in parallely
-data = X[rank * N_per_process : (rank + 1) * N_per_process]
+start_row = rank*N_per_process
+end_row = N_per_process*(rank+1)-1
+
+data = np.loadtxt(filename, delimiter=',', skiprows=start_row, max_rows=end_row-start_row+1)
```
Comment (chxru) [inline app.py:44]: Something is smelly here, you've load data in line 31 and loading again here by doing a disk IO.
I'd rather 

1. Get a total row count in CSV. You can get the total count without reading the file as mentioned in here https://stackoverflow.com/a/16108605/6862650
2. Calculate N_per_process, determine from which line to which line each process should read
3. Pass the line range, use pd.read_csv to load csv. You can set starting lines and limit by parameters there

**Review by chxru — CHANGES_REQUESTED** [review]

**Comment by sahanediriweera** [comment]
Made necessary changes 

The last discussion in PR#3 was the data was loaded again for sequential because we are supposed to take that measurement for the benchmarking

**Review by chxru — APPROVED** [review]
Formatting and further refactoring maybe required but LGTM

**Pull request merged**